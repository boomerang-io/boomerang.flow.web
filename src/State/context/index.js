import React from "react";

export const AppContext = React.createContext();

export const ExecutionContext = React.createContext({
  tasks: [],
  workflowExecution: {},
  workflowRevision: {},
});

export const EditorContext = React.createContext({
  revisionDispatch: () => {},
  revisionState: {},
  revisionQuery: {},
  summaryState: {},
  taskTemplatesData: [],
});
