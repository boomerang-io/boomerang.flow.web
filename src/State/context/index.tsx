import React from "react";
import { FlowTeam, FlowUser } from "Types";

export function createContext<ContextType>() {
  const context = React.createContext<ContextType | undefined>(undefined);
  function useContext() {
    const contextValue = React.useContext(context);
    if (!contextValue) throw new Error("useContext must be inside a Provider with a value");
    return contextValue;
  }
  return [useContext, context.Provider] as const;
}

export const [useAppContext, AppContextProvider] = createContext<AppContext>();

type AppContext = {
  isTutorialActive: boolean;
  setIsTutorialActive: (isActive: boolean) => void;
  user: FlowUser;
  teams: FlowTeam[];
};

interface taskProvider {
  category: string;
  id: string;
  icon: any;
  name: string;
}

interface stepInterface {
  activityId: string;
  duration: number;
  flowTaskStatus: string;
  id: string;
  order: number;
  outputs: any;
  startTime: string;
  taskId: string;
  taskName: string;
}

interface workflowExecutionInterface {
  creationDate: string;
  duration: number;
  id: string;
  properties: any; //array, may need to be changed later
  status: string;
  steps: Array<stepInterface>;
  teamName: string;
  trigger: string;
  workflowId: string;
  workflowRevisionid: string;
}

// export const ExecutionContext = React.createContext({});
interface executionContextInterface {
  tasks?: Array<taskProvider>;
  workflowExecution?: workflowExecutionInterface;
  workflowRevision?: object;
}
// export const ExecutionContext = React.createContext<executionContextInterface>({});
export const ExecutionContext = React.createContext<Partial<executionContextInterface>>({});

// export const EditorContext = React.createContext({});

interface revisionInterface {
  changelog: object;
  config: object;
  dag: {
    gridSize: number;
    id: string;
    links: Array<{
      color: string;
      curvyness: number;
      executionCondition: string;
      extras: object;
      id: string;
      labels: Array<string>; //i think this type is right?
      linkId: string;
      selected: false;
      source: string;
      sourcePort: string;
      switchCondition: string | null;
      target: string;
      targetPort: string;
      type: string;
      width: number;
    }>;
    nodes: Array<{
      extras: {};
      id: string;
      nodeId: string;
      passedName: string;
      ports: Array<{
        id: string;
        links: Array<string>;
        name: string;
        nodePortId: string;
        position: string;
        selected: boolean;
        type: string;
      }>;
      selected: boolean;
      templateUpgradeAvailable: boolean;
      type: string;
      x: number;
      y: number;
    }>;
    offsetX: number;
    offsetY: number;
    zoom: number;
  };
  id: string;
  templateUpgradesAvailable: string;
  version: number;
  workFlowId: string;
}

interface summaryInterface {
  data: {
    description: string;
    enableACCIntegration: boolean;
    enablePersistentStorage: boolean;
    flowTeamId: string;
    icon: string;
    id: string;
    name: string;
    properties: Array<object>;
    revisionCount: number;
    shortDescription: string;
    status: string;
    templateUpgradesAvailable: boolean;
    triggers: object;
  };
}

interface taskInterface {
  category: string;
  createdDate: string;
  currentVersion: number;
  description: string;
  icon: string;
  id: string;
  lastModified: string;
  name: string;
  nodeType: string;
  revisions: Array<object>;
  status: string;
}

interface editorContextInterface {
  revisionDispatch?: Function;
  revisionState: revisionInterface;
  summaryQuery: summaryInterface;
  taskTemplatesData: Array<taskInterface>;
}

export const EditorContext = React.createContext<Partial<editorContextInterface>>({});
