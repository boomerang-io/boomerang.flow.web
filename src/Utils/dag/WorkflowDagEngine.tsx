import { DiagramEngine, DiagramModel } from "@projectstorm/react-diagrams";
import CustomTaskNodeFactory from "./customTaskNode/CustomTaskNodeFactory";
import SimplePortFactory from "./simplePort/SimplePortFactory";
import StartEndNodeFactory from "./startEndNode/StartEndNodeFactory";
import StartEndNodeModel from "./startEndNode/StartEndNodeModel";
import StartEndPortModel from "./startEndNode/StartEndPortModel";
import SwitchLinkFactory from "./switchLink/SwitchLinkFactory";
import SwitchNodeFactory from "./switchNode/SwitchNodeFactory";
import SwitchPortModel from "./switchNode/SwitchPortModel";
import TaskLinkFactory from "./taskLink/TaskLinkFactory";
import TaskPortModel from "./taskPort/TaskPortModel";
import TemplateTaskNodeFactory from "./templateTaskNode/TemplateTaskNodeFactory";
import { NodeType } from "Constants";

interface IDagModel {
  dag: object;
  isModelLocked: boolean;
}

export default class WorkflowDagEngine {
  activeModel: DiagramModel | undefined;
  diagramEngine: DiagramEngine;

  constructor({ dag, isModelLocked }: IDagModel) {
    this.diagramEngine = new DiagramEngine();
    this.diagramEngine.installDefaultFactories();
    this.diagramEngine.registerNodeFactory(new CustomTaskNodeFactory());
    this.diagramEngine.registerNodeFactory(new StartEndNodeFactory());
    this.diagramEngine.registerNodeFactory(new TemplateTaskNodeFactory());
    this.diagramEngine.registerNodeFactory(new SwitchNodeFactory());

    //need to find a way to register port factory
    this.diagramEngine.registerPortFactory(
      new SimplePortFactory(NodeType.StartEnd, () => new StartEndPortModel({ pos: "" }))
    );
    this.diagramEngine.registerPortFactory(new SimplePortFactory(NodeType.Task, () => new TaskPortModel()));
    this.diagramEngine.registerPortFactory(
      new SimplePortFactory(NodeType.Decision, () => new SwitchPortModel({ pos: "" }))
    );

    //register new custom link
    this.diagramEngine.registerLinkFactory(new TaskLinkFactory(this.diagramEngine));
    this.diagramEngine.registerLinkFactory(new SwitchLinkFactory(this.diagramEngine));

    this.newModel(dag, isModelLocked);
  }

  newModel(dag: object, modelIsLocked: boolean) {
    this.activeModel = new DiagramModel();
    if (dag) {
      this.activeModel.deSerializeDiagram(dag, this.diagramEngine);
    } else {
      const startNode = new StartEndNodeModel({ passedName: "Start" });
      startNode.setPosition(300, 400);

      const endNode = new StartEndNodeModel({ passedName: "End" });
      endNode.setPosition(1000, 400);

      this.activeModel.addAll(startNode, endNode);
    }

    this.diagramEngine.setDiagramModel(this.activeModel);

    if (modelIsLocked) {
      this.activeModel.setLocked(true);
    }

    // Register event listeners for linkUpdated
    // We are listening on the create event and deleting it if
    // it doesn't have a target port
    this.activeModel.addListener({
      linksUpdated: (event) => {
        if (event.isCreated) {
          document.addEventListener("mouseup", (e: MouseEvent) => {
            setTimeout(() => {
              if (!event.link.targetPort) {
                this.activeModel?.removeLink(event.link);
                this.diagramEngine.repaintCanvas();
              }
            }, 0);
          });
        }
      },
    });
  }

  getActiveDiagram() {
    return this.activeModel;
  }

  getDiagramEngine() {
    return this.diagramEngine;
  }
}

export const createWorkflowRevisionBody = (
  diagramApp: WorkflowDagEngine,
  changeLogReason: string,
  workflowRevisionConfig = {}
) => {
  const dag = getDiagramSerialization(diagramApp);
  const config = formatWorkflowConfigNodes(workflowRevisionConfig);
  const changelog = {
    reason: changeLogReason,
  };
  return { dag, config, changelog };
};

const getDiagramSerialization = (diagramApp: WorkflowDagEngine): object => {
  return diagramApp.getDiagramEngine().getDiagramModel().serializeDiagram();
};

const formatWorkflowConfigNodes = (workflowRevisionConfig: object) => {
  return { nodes: Object.values(workflowRevisionConfig) };
};
