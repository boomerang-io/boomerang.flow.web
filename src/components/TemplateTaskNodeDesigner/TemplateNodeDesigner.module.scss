.closeButton {
  position: absolute;
  right: -0.75rem;
  top: -0.75rem;
  width: 1.5rem;
}

.editButton {
  position: absolute;
  right: 1.75rem;
  top: -0.75rem;
  width: 1.5rem;
}
