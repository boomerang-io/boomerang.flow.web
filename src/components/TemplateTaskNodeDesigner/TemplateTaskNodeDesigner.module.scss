.closeButton {
  position: absolute;
  right: -0.75rem;
  top: -0.75rem;
  width: 1.5rem;
}

.editButton {
  position: absolute;
  right: 1.75rem;
  top: -0.75rem;
  width: 1.5rem;
}

.updateButton {
  position: absolute;
  right: calc(50% - 0.75rem);
  top: -0.75rem;
  width: 1.5rem;
}

.updateTaskModalContainer {
  max-height: 62.5rem;
  max-width: 60rem;

  :global .bx--bmrg-modal-composed-container {
    padding-right: 25% !important;
  }
}
