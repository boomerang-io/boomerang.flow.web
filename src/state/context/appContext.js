import React from "react";

const AppContext = React.createContext({
  user: {},
  teamsQuery: {},
  teams: [],
  activeTeam: null,
  setActiveTeam: () => {},
  onBoardShow: false,
  setOnBoardShow: () => {},
});
export default AppContext;
