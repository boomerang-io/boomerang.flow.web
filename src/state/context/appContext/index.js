import React from "react";

const AppContext = React.createContext({
  user: {},
  teams: [],
  activeTeam: null,
  setActiveTeam: () => {},
  onBoardShow: false,
  setOnBoardShow: () => {},
});

export default AppContext;
