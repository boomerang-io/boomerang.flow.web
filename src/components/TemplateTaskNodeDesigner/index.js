import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { actions as workflowRevisionActions } from "State/workflowRevision";
import { actions as appActions } from "State/app";
import { ModalFlow } from "@boomerang/carbon-addons-boomerang-react";
import WorkflowCloseButton from "Components/WorkflowCloseButton";
import WorkflowEditButton from "Components/WorkflowEditButton";
import WorkflowNode from "Components/WorkflowNode";
import WorkflowTaskForm from "Components/WorkflowTaskForm";
import styles from "./TemplateNodeDesigner.module.scss";

export class TemplateNodeDesigner extends Component {
  static propTypes = {
    isModalOpen: PropTypes.bool.isRequired,
    node: PropTypes.object.isRequired,
    nodeConfig: PropTypes.object.isRequired,
    task: PropTypes.object.isRequired,
    taskNames: PropTypes.array.isRequired,
    workflowRevisionActions: PropTypes.object.isRequired
  };

  static defaultProps = {
    node: {},
    nodeConfig: {},
    task: {}
  };

  handleOnSave = inputs => {
    this.props.workflowRevisionActions.updateNodeConfig({ nodeId: this.props.node.id, inputs });
    this.forceUpdate();
  };

  // Delete the node in state and then remove it from the diagram
  handleOnDelete = () => {
    this.props.workflowRevisionActions.deleteNode({ nodeId: this.props.node.id });
    this.props.node.remove();
  };

  renderConfigureNode() {
    return (
      <ModalFlow
        confirmModalProps={{
          title: "Are you sure?",
          children: "Your changes will not be saved"
        }}
        modalHeaderProps={{
          title: `Edit ${this.props.task.name}`,
          subtitle: "Configure the inputs"
        }}
        modalTrigger={({ openModal }) => <WorkflowEditButton className={styles.editButton} onClick={openModal} />}
      >
        <WorkflowTaskForm
          inputProperties={this.props.inputProperties}
          node={this.props.node}
          nodeConfig={this.props.nodeConfig}
          onSave={this.handleOnSave}
          setIsModalOpen={this.props.appActions.setIsModalOpen}
          taskNames={this.props.taskNames}
          task={this.props.task}
        />
      </ModalFlow>
    );
  }

  render() {
    const { task, node } = this.props;
    return (
      <WorkflowNode title={task.name} subtitle={node.taskName} name={task.name} category={task.category} node={node}>
        {this.renderConfigureNode()}
        <WorkflowCloseButton className={styles.closeButton} onClick={this.handleOnDelete} />
      </WorkflowNode>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    isModalOpen: state.app.isModalOpen,
    inputProperties: state.workflow.data.properties,
    nodeConfig: state.workflowRevision.config[ownProps.node.id],
    task: state.tasks.data.find(task => task.id === ownProps.node.taskId),
    taskNames: Object.values(ownProps.diagramEngine.getDiagramModel().getNodes()) // get the taskNames names from the nodes on the model
      .map(node => node.taskName)
      .filter(name => !!name)
  };
};

const mapDispatchToProps = dispatch => ({
  appActions: bindActionCreators(appActions, dispatch),
  workflowRevisionActions: bindActionCreators(workflowRevisionActions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TemplateNodeDesigner);
