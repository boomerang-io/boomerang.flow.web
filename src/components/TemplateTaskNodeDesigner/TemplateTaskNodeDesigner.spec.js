import React from "react";
import TemplateTaskNodeDesigner from "./index";
import WorkflowDagEngine from "Utilities/dag/WorkflowDagEngine";
import { EditorContextRender } from "Utilities/testing/context";
import { revisions } from "ApiServer/fixtures";

const diagramEngine = new WorkflowDagEngine({ dag: revisions[0].dag, isLocked: false });
const node = {};

describe("Editor --- Snapshot", () => {
  it("Capturing Snapshot of Editor", () => {
    const { baseElement } = rtlContextRouterRender(
      <EditorContextRender>
        <TemplateTaskNodeDesigner diagramEngine={diagramEngine} node={node} />
      </EditorContextRender>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
