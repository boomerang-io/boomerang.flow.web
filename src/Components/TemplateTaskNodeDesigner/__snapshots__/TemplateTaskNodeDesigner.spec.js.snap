// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor --- Snapshot Capturing Snapshot of Editor 1`] = `
<body
  id="app"
>
  <div />
</body>
`;
